tls/s2n_alerts.c

```diff
@@ -315,10 +259,62 @@ S2N_RESULT s2n_queue_reader_no_renegotiation_alert(struct s2n_connection *conn)
      *# handshake_failure alert.
      **/
     if (s2n_connection_get_protocol_version(conn) == S2N_SSLv3) {
-        RESULT_GUARD_POSIX(s2n_queue_reader_alert(conn, S2N_TLS_ALERT_LEVEL_FATAL, S2N_TLS_ALERT_HANDSHAKE_FAILURE));
+        RESULT_GUARD_POSIX(s2n_queue_reader_handshake_failure_alert(conn));
+        RESULT_BAIL(S2N_ERR_BAD_MESSAGE);
+    }
+
+    if (!conn->reader_warning_out) {
+        conn->reader_warning_out = S2N_TLS_ALERT_NO_RENEGOTIATION;
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_alerts_write_error_or_close_notify(struct s2n_connection *conn)
+{
+    if (!s2n_alerts_supported(conn)) {
+        return S2N_RESULT_OK;
+    }
+
+    /* By default, s2n-tls sends a generic close_notify alert, even in
+     * response to fatal errors.
+     */
+    uint8_t code = S2N_TLS_ALERT_CLOSE_NOTIFY;
+    uint8_t level = S2N_TLS_ALERT_LEVEL_WARNING;
+
+    /* s2n-tls sends a very small subset of more specific error alerts.
+     * Since either the reader or the writer can produce one of these alerts,
+     * but only a single alert can be reported, we prioritize writer alerts.
+     */
+    if (conn->writer_alert_out) {
+        code = conn->writer_alert_out;
+        level = S2N_TLS_ALERT_LEVEL_FATAL;
+    } else if (conn->reader_alert_out) {
+        code = conn->reader_alert_out;
+        level = S2N_TLS_ALERT_LEVEL_FATAL;
+    }
+
+    struct s2n_blob alert = { 0 };
+    uint8_t alert_bytes[] = { level, code };
+    RESULT_GUARD_POSIX(s2n_blob_init(&alert, alert_bytes, sizeof(alert_bytes)));
+
+    RESULT_GUARD(s2n_record_write(conn, TLS_ALERT, &alert));
+    conn->alert_sent = true;
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_alerts_write_warning(struct s2n_connection *conn)
+{
+    if (!s2n_alerts_supported(conn)) {
         return S2N_RESULT_OK;
     }
 
-    RESULT_GUARD_POSIX(s2n_queue_reader_alert(conn, S2N_TLS_ALERT_LEVEL_WARNING, S2N_TLS_ALERT_NO_RENEGOTIATION));
+    uint8_t code = conn->reader_warning_out;
+    uint8_t level = S2N_TLS_ALERT_LEVEL_WARNING;
+
+    struct s2n_blob alert = { 0 };
+    uint8_t alert_bytes[] = { level, code };
+    RESULT_GUARD_POSIX(s2n_blob_init(&alert, alert_bytes, sizeof(alert_bytes)));
+
+    RESULT_GUARD(s2n_record_write(conn, TLS_ALERT, &alert));
```

Comment:
Do you need to set `conn->alert_sent = true` here? 🤔 I suppose you don't because it's just a warning...

Reply:
We definitely shouldn't set it here, as that would cause us not to send the alert that closes the connection. I refer to fatal and closure alerts as "alerts" (reader_alert_out, alert_sent) and warning alerts as "warnings" (reader_warning_out) in the variable naming, but maybe there's better terminology?